### ISSUE TITLE
axis differences between plot_source() and plot('source') and issues with model plots

### ISSUE THREAD

**Issue by DougBurke** [sherpa/sherpa#106]
# Summary

The data and axis labels are not the same for
- `plot_source()` compared to `plot('source')`
- `plot_model()` compared to `plot('model')`

when using PHA data sets. It's not clear whether it's the same issue for the two, but we need to look at what the `plot` command is doing.
# Details

The following is seen with the ChIPS backend (CIAO 4.7) and master with the `master` branch. I've seen it with multiple PHA data sets.

```
% ipython --pylab
Python 2.7.10 |Continuum Analytics, Inc.| (default, Sep 15 2015, 14:50:01) 
Type "copyright", "credits" or "license" for more information.

IPython 4.0.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.
Using matplotlib backend: Qt4Agg

In [1]: from sherpa.astro import ui
/home/naridge/local/anaconda/envs/sherpa-astropy-fits-gzip/lib/python2.7/site-packages/IPython/kernel/__init__.py:13: ShimWarning: The `IPython.kernel` package has been deprecated. You should import from ipykernel or jupyter_client instead.
  "You should import from ipykernel or jupyter_client instead.", ShimWarning)
WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available

In [2]: ui.load_pha(1, 'sherpa-test-data/sherpatest/threads/pha_intro/3c273.pi') 
WARNING: systematic errors were not found in file 'sherpa-test-data/sherpatest/threads/pha_intro/3c273.pi'
statistical errors were found in file 'sherpa-test-data/sherpatest/threads/pha_intro/3c273.pi' 
but not used; to use them, re-read with use_errors=True
read ARF file sherpa-test-data/sherpatest/threads/pha_intro/3c273.arf
read RMF file sherpa-test-data/sherpatest/threads/pha_intro/3c273.rmf
WARNING: systematic errors were not found in file 'sherpa-test-data/sherpatest/threads/pha_intro/3c273_bg.pi'
statistical errors were found in file 'sherpa-test-data/sherpatest/threads/pha_intro/3c273_bg.pi' 
but not used; to use them, re-read with use_errors=True
read background file sherpa-test-data/sherpatest/threads/pha_intro/3c273_bg.pi

In [3]: ui.load_pha(2, 'sherpa-test-data/sherpatest/threads/pha_intro/3c273.pi') 
WARNING: systematic errors were not found in file 'sherpa-test-data/sherpatest/threads/pha_intro/3c273.pi'
statistical errors were found in file 'sherpa-test-data/sherpatest/threads/pha_intro/3c273.pi' 
but not used; to use them, re-read with use_errors=True
read ARF file sherpa-test-data/sherpatest/threads/pha_intro/3c273.arf
read RMF file sherpa-test-data/sherpatest/threads/pha_intro/3c273.rmf
WARNING: systematic errors were not found in file 'sherpa-test-data/sherpatest/threads/pha_intro/3c273_bg.pi'
statistical errors were found in file 'sherpa-test-data/sherpatest/threads/pha_intro/3c273_bg.pi' 
but not used; to use them, re-read with use_errors=True
read background file sherpa-test-data/sherpatest/threads/pha_intro/3c273_bg.pi

In [4]: ui.set_source(1, ui.powlaw1d.pl1)

In [5]: ui.set_source(2, ui.powlaw1d.pl2)

In [6]: ui.set_analysis(1, 'energy')

In [7]: ui.set_analysis(2, 'wave')

In [8]: ui.plot_source(1)

In [9]: plt.ylim()
Out[9]: (0.0, 10.0)
```

The X axis is labelled `Energy (keV)` and the Y axis `f(E)  Photons/sec/cm^2/keV` (for the matplotlib backend this is not converted to a superscript; I'll have a PR about that).

```
In [10]: ui.plot_source(2)

In [11]: plt.ylim()
Out[11]: (0.0, 0.90000000000000002)
```

Switching to wavelength gives an X axis label of `Wavelength (\AA)` and Y axis label of `f(\lambda)  Photons/sec/cm^2/\AA`.

If I now switch to the `plot` command, I get different axis labels _and_ ranges displayed on the Y axis:

```
In [12]: ui.plot('source', 1)

In [13]: plt.ylim()
Out[13]: (0.0, 0.00040000000000000002)
```

The Y axis label is now `Counts/sec/keV`\- i.e. it looks like it thinks it is plotting the model - but you can see that the model is still a plain power law (i.e. not multiplied by an ARF). Not sure what the Yaxis range indicates, but it is _significantly_ different to the `plot_source` values shown above. You see similar things with the wavelength version:

```
In [14]: ui.plot('source', 2)

In [15]: plt.ylim()
Out[15]: (0.0, 2.9999999999999997e-05)
```

The Y axis label now is (if you can see it) `Counts/sec/Angstrom` and the X axis is `Wavelength (Angstrom)`.

if I compare `ui.plot_model(1)` to `ui.plot('model', 1)` then there's much less difference, but they aren't quite the same (it looks like the points the model is evaluated over are different, so that the X axis range is slightly different).

The plot for `ui.plot_model(2)` looks very wrong (the X axis is 0 to 5000!), whereas `ui.plot('model', 2)` covers 0 to 800.

**Comment by olaurino** [comment]
I need to translate this report in a number of tests. Here is my first attempt. @DougBurke please let me know if this makes sense or I got something wrong or something is missing. Plus, I have questions :)

So, the first (set of) test(s) is that the output of `plt.ylim()` should be the same between runs of `plot_source(n)` and `plot('source', n)` with `n=1|2`, where `set_analysis` can be set to either `wave` or `energy`. We should also test for unit strings on the `y` axis, as they should always be `Photons/sec/cm^2/\AA`. Should the values for the output of `plt.ylim()` always be `(0, 10)`?

The same should be true for `plot_model(n)` and `plot('model', n)`, however with different units (this time `Counts/sec/Angstrom`. What should the values of `plt.ylim()` be this time?

I think there is another issue in that I see `F(lambda)` rather than `F(\lambda)`.

Also, we should understand how/if we should test this with both `matplotlib` and `chips`, or maybe we can simply unit-test this making sure the backends are called with the correct parameters and data (i.e. we wouldn't test against `plt.ylim()`), but I would worry about that later.

**Comment by DougBurke** [comment]
@olaurino I made some changes to the plot labelling code after reporting this issue - see PR #107 - which explains why the report mentions labels like `f(\lambda)` and we now have `f(lambda)`. 

a) I believe that `plot_source(x)` and `plot('source', x)` should produce the same plot when there is no other change. I could imagine that - operationally - there is a reason for "small" differences between the two (they go through different code paths), but hopefully not.

Note that if I change the analysis setting then the output *will* change; that is

    set_analysis(x, 'energy')
    plot_source(x)
    set_analysis(x, 'wave')
    plot('source', x)

will produce different plots (by different I mean that for both axes the limits will be different, as will the data values plotted, and the axis labels).

b) I believe the same logic holds for the other plot types  - e.g. as you mentioned `plot_model()` and `plot('model')`.

c) Both a and b should hold for different data types (e.g. I am focussing on `DataPHA` here, but ideally we'd be testing the other plotting code). It might be easier to start with the other data types first, since they are simpler and hopefully already work!
 
d) The plot limits themselves - e.g. the output of `plt.ylim()` for matplotlib, `pychips.get_plot_yrange() for chips - depends on the data being plotted, the plot type, and other settings (primarily `set_analysis`), as well as the plot back end. That is, the plot backends are allowed to select the axis limits. I also have not looked at the data here closely enough to say what the sensible limits are.

e) The Y axis labels for the 'source' case do not always end up "/Angstrom". This depends on the `set_analysis` setting, and is a bit tricky since the optional arguments to this function - `type` and `factor` - also change the plots (so it is not just the `analysis` argument). See issue #39 for more details.

**Comment by DougBurke** [comment]
I've created the gist https://gist.github.com/DougBurke/499fd3d3d7a51e06041008c598358391 that includes an example script (essentially the example I used originally) along with the plots I get with the current master branch for 'data', 'source', and 'model'.

Note that 'source' and 'model' are expected to be different ('source' can be considered to be the spectrum that is about to enter the telescope and 'model' is what is actually detected). We also have some (probably unclear) logic on whether source and/or model is to be displayed at the native resolution or binned up to match the data ( @anetasie has a better understanding of this than I do). I mention this partly because I note that, in the figures I added to the gist, it looks like the model is being drawn at a different resolution. Compare [`out_model1_nargs1.png`](https://gist.github.com/DougBurke/499fd3d3d7a51e06041008c598358391#file-out_model1_nargs1-png) (output of `plot_model(1)`) to [`out_model1_nargs2.png`](https://gist.github.com/DougBurke/499fd3d3d7a51e06041008c598358391#file-out_model1_nargs2-png) (output of `plot('model', 1)`) and you can see that  the feature at around 3 keV is different.

**Comment by olaurino** [comment]
Alright, so this turns out to be rather tricky, and probably merits a face to face meeting to discuss what to do.

In a nutshell, the problem is that `plot` is a `sherpa.ui.utils.Session` method, inherited by `sherpa.astro.ui.utils.Session`. When you call `sherpa.astro.ui.utils.Session.plot_source`, the method checks the class of the dataset: if it's `DataPHA`, then `sherpa.ui.utils.Session._plot` is called with a special plot object defined in `sherpa.astro...Session`.

On the other hand, `sherpa.ui.utils.Session.plot` builds the call to `_plot` dynamically using the string argument passed to it. So, if you call `plot('source')`, `_plot` is given `self._sourceplot` as argument. However, when dealing with `DataPHA`s, it should be given `self._astrosourceplot`.

Right now I can't think of a way of fixing this issue without changing the fundamental architecture and make it more object oriented: if the plot depends on the data object class, then the data object class should be responsible for it and you can leverage polymorphism, without having to manually dispatch the method calls. But such change is big and risky.

The workaround is to call `plot('astrosource')` and `plot('astromodel')`, then you'll get the output you expect. But that requires the user to know when and why to use one string or the other.

**Comment by olaurino** [comment]
One way might be to override `plot` in `sherpa.astro...Session`, code the exceptions to the parent's cases, i.e. if the data is `DataPHA` and fall back on the parent's method if the data is not `DataPHA`. This might work, but it goes in the wrong direction, increasing the complexity rather than reducing it.

**Comment by DougBurke** [comment]
I did *not* know you can use `astrosource` or `astromodel` as arguments!

Thanks for looking at this; I don't think it's somehting we can reasonably fix on the 4.8.2 timescale, as I'd prefer to work on improving the API/implementation rather than adding in hacks to work around issues. However, as you said we probably need to meet to discuss this, along with the other plotting issues - #39 #106 #241 and others - to identify short, medium, and long-term targets.
